### ISSUE TITLE
volume normalisation metadata?

### ISSUE THREAD

**Issue by DeLub** [mikebrady/shairport-sync#1168]
Hi,

In my current setup I let shairport-sync play into an ALSA loopback device and have setup ffmpeg to read from the loopback device like below to equalise my headphones:

`        run_this_before_entering_active_state = "/usr/bin/ffmpeg -loglevel panic -hide_banner -nostats -nostdin -f alsa -channels 2 -sample_rate 44100 -acodec pcm_s32le -i hw:0,1 -af \"volume=volume=-4.9dB,bass=f=25:t=q:w=0.71:g=6.0,equalizer=f=150:t=q:w=0.5:g=-2.9,equalizer=f=1330...`

What I would like to do is also use ffmpeg for volume normalisation. Shairport is sending some volume metadata, but does this include normalisation data? And can I get to it? Or is this metadata not available in the stream at all, and does Apple Music perform the normalisation in the app itself before encoding it to a 16-bit alac stream (which would be quite unfortunate)?

Can you help out, please?

BTW, thank you very much for this nice piece of software.

Regards,
Alex

**Comment by mikebrady** [comment]
Thanks for the post. I'm not sure what you mean by "volume normalisation metadata", so let me explain that the audio comes in to Shairport Sync at full volume with an attendant "AirPlay Volume" of between `-30.00` and `0.0`, which seems to be logarithmic. A further value that may appear is `-144.0`, indicating `mute`.

The unmodified audio and the AirPlay volume setting can both be available to you from Shairport Sync. The AirPlay volume is available in the metadata pipe or on the `run_this_when_volume_is_set` hook as a parameter. Unmodified audio is available if you set the `ignore_volume_control` to `"yes"`.

**Comment by DeLub** [comment]
Sorry for not being clear.

What I meant is: I know that some metadata of the track currently playing is sent over to Shairport. `volumeNormalizationEnergy` is also part of the metadata in iTunes/Apple Music. It is used to perform volume normalisation. ([info on Apple Developer Documentation](https://developer.apple.com/documentation/ituneslibrary/itlibmediaitem/1811123-volumenormalizationenergy)) My question is: is this metadata also transferred via the AirPlay protocol? If so, this would open up opportunities to do volume normalisation on the receiver's end. I would prefer this to having it done by Apple Music. Or is the Apple Music player "baking" this in and actually changing the bits that are being sent over? You're update seems to suggest this.

**Comment by mikebrady** [comment]
Thanks for the clarification. It may well be that the volume normalization energy data is being passed over in the metadata -- I simply don't know; quite a bit of the metadata is unrecognised and discarded -- please see the code in the [Shairport Sync Metadata Reader](https://github.com/mikebrady/shairport-sync-metadata-reader).

What I meant to emphasise in the update you refer to was that you can turn off any modifications that Apple "bake in" to the audio data. So, if the volume normalization energy is indeed being passed in, you could have the original audio and the volume normalization energy information as you need.

**Comment by DeLub** [comment]
This is great! I found out that the metadata is indeed included (code 'aeNV'). The type according to the ytrack - DMAP.wiki is 5, long, 4 byte integer. I'm not sure how to decode this payload, though. The default decoding function in your metadata reader provides ridiculously large numbers, but it is probably specifically meant to decode strings. Can you give me some pointer on how this should be decoded, please?

**Comment by mikebrady** [comment]
It's seems to be a 32-bit integer. If you look at the code, you'll see where a 64-bit integer is decoded by treating it as two 32-bit integers, so you could adapt that code.

**Comment by DeLub** [comment]
I tried that but just made a stupid mistake... Anyway, I got the values now.

So: the value is an integer between 1 and ...? 0 signifies that the normalisation value is unknown. Otherwise the required gain change in dB can be calculated by -10*log(value)+30. This brings the level up to -16dBFS. This formula matches with a random sample of 10 songs from my library.

I think it would be possible to take this number from the metadata and perform the normalisation in shairport-sync. However... This value is always provided on a track-by-track basis. This is great for playlists, but not so for albums, especially classical music. For albums we need to apply the same normalisation factor to all songs on the album. This value seems not to be present in the metadata.

Apple Music (on the desktop) does perform normalisation on album level, but I have no clue how. Also: it too often raises the volume when listening to classical music. I would rather normalise to say -23dBFS which was one of the reasons to look into this. Another is that Music on iOS can only do track-by-track normalisation (this is a bug IMO).

I seem to be stuck... not sure if there is a way to get 'unstuck'.
